File: src/types/src/agents/src/agents/workspace.agent.ts
import type { AgentSignal, UserStateProfile } from "../types/adaptive";

export function runWorkspaceAgent(state: UserStateProfile): AgentSignal[] {
  const signals: AgentSignal[] = [];

  // Check if user is ready to unlock next tier
  const canUnlockTier2 = state.tier === 1 && state.recentWinsCount >= 1;
  const canUnlockTier3 = state.tier === 2 && state.recentWinsCount >= 3;
  const canUnlockTier4 = state.tier === 3;

  if (canUnlockTier2) {
    signals.push({
      agent: "workspace",
      signal: "user_ready_for_unlock",
      confidence: 0.95,
      summary: "User has logged their first win. Daily Rituals and Wins are ready to unlock.",
      suggestedAction: "surface_unlock_prompt",
      payload: { nextTier: 2 },
    });
  } else if (canUnlockTier3) {
    signals.push({
      agent: "workspace",
      signal: "user_ready_for_unlock",
      confidence: 0.95,
      summary: "3+ wins logged. Flow Priming, Energy Patterns, and Priorities are ready to unlock.",
      suggestedAction: "surface_unlock_prompt",
      payload: { nextTier: 3 },
    });
  } else if (canUnlockTier4) {
    signals.push({
      agent: "workspace",
      signal: "user_ready_for_unlock",
      confidence: 0.7,
      summary: "User has been using the app consistently. Full toolkit is available on request.",
      suggestedAction: "surface_unlock_prompt",
      payload: { nextTier: 4 },
    });
  }

  // Workspace complexity vs tolerance
  if (state.overwhelmRisk > 0.5 && state.tier >= 3) {
    signals.push({
      agent: "workspace",
      signal: "workspace_too_complex",
      confidence: state.overwhelmRisk,
      summary: "High overwhelm risk with a complex workspace. Consider temporarily simplifying the view.",
      suggestedAction: "simplify_workspace",
    });
  }

  // Healthy workspace
  if (
    state.tier >= 2 &&
    state.overwhelmRisk < 0.3 &&
    state.momentumScore > 0.4 &&
    state.workspaceComplexityTolerance > 0.5
  ) {
    signals.push({
      agent: "workspace",
      signal: "workspace_healthy",
      confidence: 0.8,
      summary: "Workspace complexity matches user tolerance. Good balance.",
    });
  }

  // Underused modules
  const underused = state.activeModules.filter(m => (state.moduleUsage[m] ?? 0) < 2);
  if (underused.length > 2) {
    signals.push({
      agent: "workspace",
      signal: "hide_unused_module",
      confidence: 0.6,
      summary: `Several active modules have low usage: ${underused.join(", ")}. Could declutter the sidebar.`,
      payload: { modules: underused },
    });
  }

  return signals;
}

File: src/types/src/agents/python/src/agents/python/src/agents/python/src/agents/python/workspace_agent.py
from typing import List
from types import UserStateProfile, AgentSignal

def run_workspace_agent(state: UserStateProfile) -> List[AgentSignal]:
    """
    Workspace Agent - Manages feature unlocks, UI complexity, and workspace health.
    
    Watches:
    - Feature tier progression
    - Overwhelm risk vs complexity tolerance
    - Module usage patterns
    """
    signals: List[AgentSignal] = []

    # Tier unlock detection
    can_unlock_tier_2 = state.tier == 1 and state.recent_wins_count >= 1
    can_unlock_tier_3 = state.tier == 2 and state.recent_wins_count >= 3
    can_unlock_tier_4 = state.tier == 3

    if can_unlock_tier_2:
        signals.append(AgentSignal(
            agent="workspace",
            signal="user_ready_for_unlock",
            confidence=0.95,
            summary="User has logged their first win. Daily Rituals and Wins are ready to unlock.",
            suggested_action="surface_unlock_prompt",
            payload={"next_tier": 2}
        ))
    
    elif can_unlock_tier_3:
        signals.append(AgentSignal(
            agent="workspace",
            signal="user_ready_for_unlock",
            confidence=0.95,
            summary="3+ wins logged. Flow Priming, Energy Patterns, and Priorities are ready to unlock.",
            suggested_action="surface_unlock_prompt",
            payload={"next_tier": 3}
        ))
    
    elif can_unlock_tier_4:
        signals.append(AgentSignal(
            agent="workspace",
            signal="user_ready_for_unlock",
            confidence=0.7,
            summary="User has been using the app consistently. Full toolkit is available on request.",
            suggested_action="surface_unlock_prompt",
            payload={"next_tier": 4}
        ))

    # Workspace complexity check
    if state.overwhelm_risk > 0.5 and state.tier >= 3:
        signals.append(AgentSignal(
            agent="workspace",
            signal="workspace_too_complex",
            confidence=state.overwhelm_risk,
            summary="High overwhelm risk with a complex workspace. Consider temporarily simplifying the view.",
            suggested_action="simplify_workspace"
        ))

    # Healthy workspace indicator
    if (state.tier >= 2 and
        state.overwhelm_risk < 0.3 and
        state.momentum_score > 0.4 and
        state.workspace_complexity_tolerance > 0.5):
        signals.append(AgentSignal(
            agent="workspace",
            signal="workspace_healthy",
            confidence=0.8,
            summary="Workspace complexity matches user tolerance. Good balance.",
            suggested_action=None
        ))

    # Underused modules detection
    underused = [
        module for module in state.active_modules
        if state.module_usage.get(module, 0) < 2
    ]
    
    if len(underused) > 2:
        modules_str = ", ".join(underused)
        signals.append(AgentSignal(
            agent="workspace",
            signal="hide_unused_module",
            confidence=0.6,
            summary=f"Several active modules have low usage: {modules_str}. Could declutter the sidebar.",
            payload={"modules": underused}
        ))

    return signals
